### ISSUE TITLE
Dropout needed in inference mode (Pytorch)

### ISSUE THREAD

**Issue by ArchieLuxton** [onnx/onnx#4855]
Following on from issue 3753 ([now closed](https://github.com/onnx/onnx/issues/3753#issue-1013881576)):

My Pytorch model uses dropout in inference mode as well as training, but I'm struggling to pull this off when exporting to onnx. 

My model is defined as:

```
class MC_Dropout_Model(nn.Module):
    """
    A class to define the network implementing Dropout.
    Inference is performed later by a MC sampling method.
    """

    def __init__(self, input_dim, output_dim,
                 num_units, drop_prob, num_layers):
        super(MC_Dropout_Model, self).__init__()

        # input and output dimensions
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.drop_prob = drop_prob
        self.num_layers = num_layers

        # n hidden layers
        layers = OrderedDict()

        # Input layer
        layers["0"] = nn.Linear(input_dim, num_units)

        for i in range(1, num_layers - 1):
            layers[str(i)] = nn.Linear(num_units, num_units)

        # Output layer
        layers[str(num_layers - 1)] = nn.Linear(num_units, output_dim)

        self.layers = nn.Sequential(layers)

    def forward(self, x):
        x = x.view(-1, self.input_dim)
        for layer in self.layers[:-1]:
            x = layer(x)
            x = torch.relu(x)
            x = F.dropout(x, p=self.drop_prob, training=True)
        x = self.layers[self.num_layers - 1](x)  # OUTPUT LAYER
        return x
```

And I'm using the following to export and check the model:

 ```
  # Export the model
  torch.onnx.export(torch_model,               # model being run
                    x,                         # model input (or a tuple for multiple inputs)
                    "mc_dropout.onnx",   # where to save the model (can be a file or file-like object)
                    training=torch.onnx.TrainingMode.TRAINING,
                    export_params=True,        # store the trained parameter weights inside the model file
                    opset_version=10,          # the ONNX version to export the model to
                    do_constant_folding=True,  # whether to execute constant folding for optimization
                    input_names = ["input"],   # the model's input names
                    output_names = ["output"], # the model's output names
                    dynamic_axes={"input" : {0 : "batch_size"},    # variable length axes
                                  "output" : {0 : "batch_size"}})
  
  # Load the ONNX model
  model = onnx.load("mc_dropout.onnx")
  
  # Check that the IR is well formed
  onnx.checker.check_model(model)
  
  # Print a human readable representation of the graph
  print(onnx.helper.printable_graph(model.graph))
```

Which then prints the following:

```
  graph torch_jit (
    %input[FLOAT, batch_size]
  ) initializers (
    %layers.0.weight[FLOAT, 50x4]
    %layers.0.bias[FLOAT, 50]
    %layers.1.weight[FLOAT, 50x50]
    %layers.1.bias[FLOAT, 50]
    %layers.2.weight[FLOAT, 50x50]
    %layers.2.bias[FLOAT, 50]
    %layers.3.weight[FLOAT, 1x50]
    %layers.3.bias[FLOAT, 1]
  ) {
    %onnx::Reshape_9 = Constant[value = <Tensor>]()
    %onnx::Gemm_10 = Reshape(%input, %onnx::Reshape_9)
    %onnx::Relu_11 = Gemm[alpha = 1, beta = 1, transB = 1](%onnx::Gemm_10, %layers.0.weight, %layers.0.bias)
    %x = Relu(%onnx::Relu_11)
    %onnx::Gemm_13, %14 = Dropout[ratio = 0.349999994039536](%x)
    %onnx::Relu_15 = Gemm[alpha = 1, beta = 1, transB = 1](%onnx::Gemm_13, %layers.1.weight, %layers.1.bias)
    %x.3 = Relu(%onnx::Relu_15)
    %onnx::Gemm_17, %18 = Dropout[ratio = 0.349999994039536](%x.3)
    %onnx::Relu_19 = Gemm[alpha = 1, beta = 1, transB = 1](%onnx::Gemm_17, %layers.2.weight, %layers.2.bias)
    %x.7 = Relu(%onnx::Relu_19)
    %onnx::Gemm_21, %22 = Dropout[ratio = 0.349999994039536](%x.7)
    %output = Gemm[alpha = 1, beta = 1, transB = 1](%onnx::Gemm_21, %layers.3.weight, %layers.3.bias)
    return %output
  }
```

Then using the following to perform inference. Unfortunately at the moment, every forward pass returns the same value, whereas the dropout should be applied giving a different result each time. 

```
sess_options = onnxruntime.SessionOptions()

# Set graph optimization level
sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL
ort_session = onnxruntime.InferenceSession("mc_dropout.onnx", disabled_optimizers=["EliminateDropout"])

onnxruntime.set_seed(np.random.randint(123624734))

def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

# compute ONNX Runtime output prediction
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
ort_outs = ort_session.run(None, ort_inputs)
```

Any advice would be greatly appreciated.
Thanks

**Comment by jcwchen** [comment]
Hi @ArchieLuxton,
It's more a question for ONNX Runtime. Have you tried this approach https://github.com/onnx/onnx/issues/3753#issuecomment-1150691101? If it still not helps, I would suggest you raise issue in ONNX Runtime to get help from Runtime experts.

**Comment by ArchieLuxton** [comment]
Hey @jcwchen , yep I tried that with no success unfortunately.
